src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> firstNode;
+    private Node<T> lastNode;
+    private int size;
+
+    public MyLinkedList() {
+        Node<T> firstNode = null;
+        Node<T> lastNode = null;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<T>(lastNode, value, null);
+        if (firstNode == null) {
+            firstNode = node;
+            lastNode = node;
+            size++;
+            return;
+        }
+        if (firstNode.next == null) {
+            firstNode.next = node;
+        }
+        lastNode.next = node;
+        lastNode = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndexInBounds(index);
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                Node<T> node = new Node<T>(currentNode.prev, value, currentNode);
+                if (index == 0) {
+                    firstNode = node;
+                } else {
+                    currentNode.prev.next = node;
+                }
+                currentNode.prev = node;
+                size++;
+                return;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexInBounds(index);
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                return currentNode.value;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
+        checkIndexInBounds(index);
+        Node<T> currentNode = firstNode;
+        int i = 0;
+        while (currentNode != null) {
+            if (i == index) {
+                T oldValue = currentNode.value;
+                currentNode.value = value;
+                return oldValue;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
It can be made a little easier.
Get the node by index using the method written earlier.
Then make unlink (for this, it is also better to make a separate method, because the logic is repeated).